File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/Complex/Table/Columns.cs
using System.Collections.Generic;
using System.Linq;
using JDI_Commons;
using JDI_Web.Selenium.Elements.Base;
using JDI_Web.Selenium.Elements.Complex.Table.Interfaces;
using OpenQA.Selenium;
using static Epam.JDI.Core.Settings.JDISettings;
using static Epam.JDI.Core.ExceptionUtils;

namespace JDI_Web.Selenium.Elements.Complex.Table
{
    public class Columns : TableLine
    {
        public Columns()
        {
            HasHeader = true;
            ElementIndex = ElementIndexType.Nums;
            HeadersLocator = By.XPath(".//th");
            DefaultTemplate = By.XPath(".//tr/td[{0}]");
        }

        protected List<IWebElement> GetHeadersAction()
        {
            return Table.WebElements.FindAll(el => el.Equals(HeadersLocator));
        }

        protected override IList<IWebElement> GetFirstLine => Table.Rows.GetLineAction(1);
        
        public Dictionary<string, ICell> GetColumn(string colName)
        {
            return ActionWithException(() =>
            {
                var rowsCount = Table.Rows.Count;
                var webColumn = Timer.GetResultByCondition(() => GetLineAction(colName), els => els.Count >= rowsCount);
                if (webColumn == null)
                    throw Exception($"Table has only {GetLineAction(colName).Count} columns " +
                                    $"but expected at least {rowsCount}");

                var result = new Dictionary<string, ICell>();
                if (webColumn.Count == rowsCount)
                {
                    AddRows(result, Table.Rows.Headers, webColumn, colName);
                    return result;
                }
                AddRows(result, Table.Rows.AllHeaders, webColumn, colName);
                return result.Where(el => Table.Rows.Headers.Contains(el.Key)).ToDictionary();
            }, ex => $"Can't Get Column '{colName}'. Reason: {ex}");
        }

        private void AddRows(Dictionary<string, ICell> result, IList<string> headers, IList<IWebElement> webColumn, string colName)
        {
            for (var i = 0; i < headers.Count; i++)
            {
                var label = headers[i];
                if (!result.ContainsKey(label))
                    result.Add(label, Table.Cell(webColumn[i], new Column(name: colName), new Row(name: label)));
            }
        }

        public IList<string> GetColumnValue(string colName)
        {
            return
                ActionWithException(
                    () => GetLineAction(colName).Select(el => el.Text).ToList(), 
                    ex => $"Can't Get Column '{colName}'. Reason: {ex}");
        }

        public Dictionary<string, string> GetColumnAsText(string colName)
        {
            return GetColumn(colName).ToDictionary(pair => pair.Key, pair => pair.Value.GetText);
        }

        public Dictionary<string, ICell> CellsToColumn(IEnumerable<ICell> cells)
        {
            return cells.ToDictionary(cell => Headers[cell.RowNum - 1], cell => cell);
        }

        public Dictionary<string, ICell> GetColumn(int colNum)
        {
            if (Count < 0 || Count < colNum || colNum <= 0)
                throw Exception($"Can't Get Column '{colNum}'. [num] > RowsCount({Count}).");
            return ActionWithException(() =>
            {
                var rowsCount = Table.Rows.Count;
                var webColumn = Timer.GetResultByCondition(() => GetLineAction(colNum),
                    els => els.Count >= rowsCount);
                if (webColumn == null)
                    throw Exception($"Table has only {GetLineAction(colNum).Count} columns " +
                                    $"but expected at least {rowsCount}");
                var result = new Dictionary<string, ICell>();
                if (webColumn.Count == rowsCount)
                {
                    AddRows(result, Table.Rows.Headers, webColumn, colNum);
                    return result;
                }
                AddRows(result, Table.Rows.AllHeaders, webColumn, colNum);
                return result.Where(el => Table.Rows.Headers.Contains(el.Key)).ToDictionary();
            }, ex => $"Can't Get Column '{colNum}'. Reason: {ex}");
        }

        /*private Dictionary<string, Dictionary<string, ICell>> WithValueByRule(
            Row row, JFuncTTREx<string, string, bool> func)
        {
            List<string> rowNames = row.HasName
                ? Table.Rows.GetRowAsText(row.Name).where(func).keys()
                : Table.Rows.GetRowAsText(row.Name).where(func).keys();
            return new Dictionary<>(rowNames, key=>key, this::getColumn);
        }*/

        private void AddRows(Dictionary<string, ICell> result, IList<string> headers, IList<IWebElement> webColumn, int colNum)
        {
            for (var i = 0; i < headers.Count; i++)
                result.Add(headers[i], Table.Cell(webColumn[i], new Column(colNum), new Row(i + 1)));
        }

        public IList<string> GetColumnValue(int colNum)
        {
            if (Count < 0 || Count < colNum || colNum <= 0)
                throw Exception($"Can't Get Column '{colNum}'. [num] > RowsCount({Count}).");
            return ActionWithException(() => GetLineAction(colNum).Select(el => el.Text).ToList(),
                ex => $"Can't Get Column '{colNum}'. Reason: {ex}");
        }

        public Dictionary<string, string> GetColumnAsText(int colNum)
        {
            return GetColumn(colNum).ToDictionary(pair => pair.Key, pair => pair.Value.GetText);
        }

        public override Dictionary<string, Dictionary<string, ICell>> Get()
        {
            return Headers.ToDictionary(key =>key, GetColumn);
        }
        
    }
}

File: jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winnium/elements/complex/table/Columns.java
package com.epam.jdi.uitests.win.winnium.elements.complex.table;

import com.epam.commons.linqinterfaces.JFuncTTREx;
import com.epam.commons.map.MapArray;
import com.epam.jdi.uitests.core.interfaces.common.IText;
import com.epam.jdi.uitests.core.interfaces.complex.interfaces.Column;
import com.epam.jdi.uitests.core.interfaces.complex.interfaces.ICell;
import com.epam.jdi.uitests.core.interfaces.complex.interfaces.IColumn;
import com.epam.jdi.uitests.core.interfaces.complex.interfaces.Row;
import org.openqa.selenium.WebElement;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

public class Columns extends TableLine implements IColumn {
    public Columns(Table table) {
        super(table);
    }

    @Override
    public MapArray<String, ICell> getColumn(String colName) {
        int rowsCount = table.rows().count();

        return new MapArray<>(rowsCount, key -> table.rows().headers().get(key),
                value -> table.cell(new Column(colName), new Row(value + 1)));
    }

    @Override
    public List<String> getColumnValue(String colName) {
        return getLineAction(colName).stream().map(WebElement::getText).collect(Collectors.toList());
    }

    @Override
    public MapArray<String, String> getColumnAsText(String colName) {
        return getColumn(colName).toMapArray(IText::getText);
    }

    private MapArray<String, MapArray<String, ICell>> withValueByRule(Row row,
                                                                      JFuncTTREx<String, String, Boolean> func) {
        Collection<String> rowNames = row.hasName()
                ? table.rows().getRowAsText(row.getName()).where(func).keys()
                : table.rows().getRowAsText(row.getNum()).where(func).keys();
        return new MapArray<>(rowNames, key -> key, this::getColumn);
    }

    @Override
    public final MapArray<String, MapArray<String, ICell>> withValue(String value, Row row) {
        return withValueByRule(row, (key, val) -> val.equals(value));
    }

    @Override
    public final MapArray<String, MapArray<String, ICell>> containsValue(String value, Row row) {
        return withValueByRule(row, (key, val) -> val.contains(value));
    }

    @Override
    public final MapArray<String, MapArray<String, ICell>> matchesRegEx(String regEx, Row row) {
        return withValueByRule(row, (key, val) -> val.matches(regEx));
    }

    @Override
    public MapArray<String, ICell> getColumn(int colNum) {
        if (count() < 0 || count() < colNum || colNum <= 0)
            throw exception("Can't Get Column '%s'. [num] > RowsCount(%s).", colNum, count());

        int rowsCount = table.rows().count();

        return new MapArray<>(rowsCount, key -> table.rows().headers().get(key),
                value -> table.cell(new Column(colNum), new Row(value + 1)));
    }

    @Override
    public List<String> getColumnValue(int colNum) {
        if (count() < 0 || count() < colNum || colNum <= 0)
            throw exception("Can't Get Column '%s'. [num] > RowsCount(%s).", colNum, count());

        return getLineAction(colNum).stream().map(WebElement::getText).collect(Collectors.toList());
    }

    @Override
    public MapArray<String, String> getColumnAsText(int colNum) {
        return getColumn(colNum).toMapArray(IText::getText);
    }

    @Override
    protected List<WebElement> getFirstLine() {
        return ((Rows)table.rows()).getLineAction(1);
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> get() {
        return new MapArray<>(headers(), key -> key, this::getColumn);
    }
}
